### PR TITLE
8313181: Enabling modern media controls on webkit 616.1 does not load button images on HTML5 video Element

### PR DESCRIPTION
**Pull request by jaybhaskar**
Issue : Enabling modern media controls on webkit 616.1 does not load button images on HTML5 Video Element
Solution: Add resources and correct MediaControl Stylesheet

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8313181](https://bugs.openjdk.org/browse/JDK-8313181): Enabling modern media controls on webkit 616.1 does not load button images on HTML5 video Element (**Bug** - P3)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Hima Bindu Meda](https://openjdk.org/census#hmeda) (@HimaBinduMeda - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1201/head:pull/1201` \
`$ git checkout pull/1201`

Update a local copy of the PR: \
`$ git checkout pull/1201` \
`$ git pull https://git.openjdk.org/jfx.git pull/1201/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1201`

View PR using the GUI difftool: \
`$ git pr show -t 1201`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1201.diff">https://git.openjdk.org/jfx/pull/1201.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1201#issuecomment-1666831638)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
Reviewers: @kevinrushforth @HimaBinduMeda 

/reviewers 2

File: tests/manual/web/HTML5VideoControlTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        VBox instructionsBox = new VBox();
+        instructionsBox.setStyle("-fx-background-color: #f0f0f0;");
+        instructionsBox.setPrefWidth(400);
+
+        instructionsBox.getChildren().addAll(
+            new javafx.scene.control.Label("Instructions:"),
+            new javafx.scene.control.Label("1. Click 'Play' to start the video."),
+            new javafx.scene.control.Label("2. Use the video controls to pause,"),
+            new javafx.scene.control.Label("2. controls should be visible,"),
+            new javafx.scene.control.Label("4. it works!")
```
Comment (kevinrushforth) [inline tests/manual/web/HTML5VideoControlTest.java:68]: You have two step 2 lines, and step 4 isn't precise enough. I would recommend replacing both line 2s and line 4 with something like:

```
    2. The media controls should be visible once the video starts playing.
    3. Use the media controls to play/pause/seek the video.
```

File: modules/javafx.web/src/main/native/Source/WebCore/platform/java/generate_icon_resource.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+'''
+'''
+Tool to generate base64 code
```
Comment (kevinrushforth) [inline modules/javafx.web/src/main/native/Source/WebCore/platform/java/generate_icon_resource.py:30]: There is a trailing space on this line. We don't currently enforce whitespace rules for `.py` files, but might in the future.

**Review by kevinrushforth — COMMENTED** [review]
I tested this on Windows and confirm that the media controls are visible and active now. I left a few inline comments on the test and the utility script.

File: modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+
```
[the remark this answers is not in the document]
Comment (jaybhaskar) [reply modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp]: done

File: modules/javafx.web/src/main/native/Source/WebCore/platform/java/generate_icon_resource.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (jaybhaskar) [reply modules/javafx.web/src/main/native/Source/WebCore/platform/java/generate_icon_resource.py]: The python would be used to generate base64 code of images and will generate ModernMediaControlResource.cpp at build time later

File: tests/manual/web/HTML5VideoControlTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        primaryStage.show();
+    }
+
+    private VBox createInstructionsBox() {
+        VBox instructionsBox = new VBox();
+        instructionsBox.setStyle("-fx-background-color: #f0f0f0;");
+        instructionsBox.setPrefWidth(400);
+
+        instructionsBox.getChildren().addAll(
+            new javafx.scene.control.Label("Instructions:"),
```
[the remark this answers is not in the document]
Comment (jaybhaskar) [reply tests/manual/web/HTML5VideoControlTest.java]: done

File: tests/manual/web/HTML5VideoControlTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import javafx.scene.web.WebView;
+import javafx.stage.Stage;
+
+public class HTML5VideoControlTest extends Application {
+    @Override
+    public void start(Stage primaryStage) {
+        WebView webView = new WebView();
+        webView.getEngine().loadContent(
+            "<video width=\"320\" height=\"240\" controls>\n" +
+            "  <source src=\"your_video_url.mp4\" type=\"video/mp4\">\n" +
```
[the remark this answers is not in the document]
Comment (jaybhaskar) [reply tests/manual/web/HTML5VideoControlTest.java]: done

File: tests/manual/web/HTML5VideoControlTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            "  Your browser does not support the video tag.\n" +
+            "</video>"
+        );
+
+        VBox root = new VBox();
+        root.setSpacing(20);
+
+        HBox content = new HBox();
+        content.getChildren().addAll(createInstructionsBox(), webView); // Swap the order here
```
[the remark this answers is not in the document]
Comment (jaybhaskar) [reply tests/manual/web/HTML5VideoControlTest.java:49]: it is follow up

File: modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "ModernMediaControlResource.h"
+
+ModernMediaControlResource::ModernMediaControlResource() {
```
Comment (kevinrushforth) [inline modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp:28]: Can you add a comment indicating that the Base64 images in this file were generated from the resources in `WebCore/Modules/modern-media-controls/images/adwaita/` using the `Tools/Scripts/generate_icon_resource.py` script?

File: tests/manual/web/HTML5VideoControlTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply tests/manual/web/HTML5VideoControlTest.java:49]: But why? What is the reason you want to make the change? I don't see it as needed.

More to the point, it is probably better to add the instructions to the root `VBox` either above or below the WebView. As it is, the instructions are truncated. See the attached image.

**Review by kevinrushforth — COMMENTED** [review]
The updates look good with two remaining comments.

Here is the screen shot that shows the clipping in the example test program.

HTML5VideoControlTest

File: modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp (hunk 8 above)
Comment (jaybhaskar) [reply modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp:28]: done

Comment (jaybhaskar) [reply modules/javafx.web/src/main/native/Source/WebCore/platform/java/ModernMediaControlResource.cpp:28]: done

**Review by kevinrushforth — APPROVED** [review]

**Review by HimaBinduMeda — APPROVED** [review]
Changes look fine. Internal testing looks fine.

**Pull request closed without merging**